api/envoy/api/v2/listener/listener.proto

```diff
@@ -47,12 +47,28 @@ message Filter {
 
 // Specifies the match criteria for selecting a specific filter chain for a
 // listener.
+//
+// In order for a filter chain to be selected, *ALL* of its criteria must be
+// fulfilled by the incoming connection, properties of which are set by the
+// networking stack and/or listener filters.
+//
+// The following order applies:
+//
+// 1. Server name (e.g. SNI for TLS protocol),
+// 2. Transport protocol.
+//
+// For criterias that allow ranges or wildcards, the most specific value in any
+// of the configured filter chains that matches the incoming connection is going
+// to be used (e.g. for SNI ``www.example.com`` the most specific match would be
+// ``www.example.com``, then ``*.example.com``, then any filter chain without
+// ``sni_domains`` requirements).
```

Comment:
What about when we have CIDR and SNI, what order do we resolve with? In general, it would be helpful to clarify where we want to go in the end state of the API for all the match criteria and precedence.

Reply:
Sure, I can see this being a useful thing to add later. So, for now, as long as we have precedence for the fields that are actually implemented, and a TODO for the rest I'd be happy.

Reply:
As mentioned earlier (I think?), the destination IP would take precedence over SNI, over transport protocol. I didn't consider source IPs.

I could see having `filter_chain_match_precedence: ["server_names", "source_ranges", "transport_protocol"]` as a way of controlling precedence, but it doesn't seem that this is something required by most of the use cases, so I'm going to let someone that has an use case for it work on this.

As for the "end goal" docs - I'm under the impression that the API and docs document current state, and not future vision, so I don't think the "end goal" should be added here, especially since things change during implementation, and I wouldn't want to promise stuff that won't be eventually implemented.

Reply:
This isn't quite correct. The user facing API docs should reflect the current state. We do explicit hidden TODOs in various places to signal to developers and those interested in understanding future directions where we want to head, e.g. https://github.com/envoyproxy/envoy/blob/6ae5612ddb84592a617d2b6b423e3d8c5d2081bc/api/envoy/config/accesslog/v2/als.proto#L13.

Please do add appropriate TODOs in that style and preferably open an issue to track future work here. That's all that's being asked at this point.